src/box/memtx_tree.cc

```diff
@@ -886,10 +913,26 @@ memtx_tree_lookup(memtx_tree_t<USE_HINT> *tree,
 		 * the tree are less (less or equal) to the key, and
 		 * iterator_prev call will convert the iterator to the
 		 * last position in the tree, that's what we need.
+		 *
+		 * We only step back if we don't need to skip tuples,
+		 * otherwise we will only need the current offset to
+		 * update the iterator.
 		 */
-		memtx_tree_iterator_prev(tree, iterator);
+		if (skip == 0)
+			memtx_tree_iterator_prev(tree, iterator);
 		--*offset; /* Unsigned underflow possible. */
 	}
+
+	/*
+	 * Skip the amount of tuples required. Here raw tuples are skipped,
+	 * but in case of multiple concurrent transactions, some of these
+	 * tuples might be invisible to the current one, this is handled on
+	 * the caller site.
+	 */
+	if (skip != 0) {
+		memtx_tree_skip(tree, iterator, offset, skip,
+				iterator_type_is_reverse(*type));
+	}
```

Comment:
IMO better do it outside `memtx_tree_lookup` - in `tree_iterator_start`. Also, I'd inline `memtx_tree_skip` since it's small, used in one place, and having it closer to the rest of the "skip" logic would make the code easier to follow.

Reply:
This function can't be used in the read view iterator as is because we use different types for read views. I guess we'll have to duplicate the code anyway.

Reply:
This is a matter of overloading a bunch of functions: see the [corresponding commit in the read-view PR](https://github.com/tarantool/tarantool/pull/10478/commits/a83f5e3c42aaf27ebfdd8b3850b9df475b48eb34).

This may make code bigger, but it creates a single point of responsibility for the lookup behavior.

The same trick is done for the read-view [count implementation](https://github.com/tarantool/tarantool/pull/10357/commits/9316f0643992f428d47c216b212f83550240f7f3), so it's generic too.

Reply:
I don't think it's a good idea because tree and its view are handled differently. For example, we don't need to do any kind of tracking in a view in contrast to a tree. If the tx manager was implemented at a different level (not in the index), it could make sense but otherwise it will probably make the code convoluted IMO. Besides, the view is implemented in EE and it should stay there for now.

Reply:
Inlined the iterator skip into the `tree_iterator_start`.